### ISSUE TITLE
What the motivation of this fork?

### ISSUE THREAD

**Issue by JounQin** [es-tooling/eslint-plugin-unicorn-x#6]
Is there any related discussion contexts?

**Comment by lishaduck** [comment]
Nobody's watching this repo, so I assume @43081j is unaware this was raised.
I'm also curious, so hopefully that pinged him.

**Comment by JounQin** [comment]
@lishaduck See https://discord.com/channels/1227627367204257852/1365520775637106788

**Comment by lishaduck** [comment]
Can you paste in a screenshot or whatever for those who don't use dis?

**Comment by 43081j** [comment]
Sorry folks, I'll write up some reasoning today

Importantly, this hasn't been published yet and some of it will probably be used to open upstream PRs later

Whether it will exist as it's own fork yet is to be decided (but possible).

**Comment by 43081j** [comment]
ok here you go:

easy wins:

- `empathic` is faster and smaller than `read-package-up` (i hear you're moving to `find-up-simple`, but this will still be true)
- `is-builtin-module` is unnecessary, node ships a built in list now (we have to exclude `punycode` since the npm package didn't include it)
- `indent-string` is unnecessary, especially given it is used to apply a `count` of indentation but we _always_ call it with `1` and set the indent _character_ to the entire indent string (i.e. the already `repeat`-ed indent character)
- `strip-indent` can be replaced with `dedent` (will be about as big when it switches to ESM-only)
- `tinyexec` de-dupes better than `nano-spawn`

more involved wins:

- a custom build of lodash is shipped pretty much just to get some string casing utils - replace this with scule, smaller and faster
- eslint doc generator no longer needs the hack (`workaround-for-eslint-doc-generator`)

most of this can be contributed back upstream, but i suspect are unlikely to ever be accepted.

i then got carried away and diverged the fork to do a whole bunch of other stuff the original can't feasibly ever really pull in:

- overhaul the dev setup (just rip everything out and use vitest, prettier, simplify lint config, etc)

on top of that, i did some quick profiling of the rules and found many places we can probably speed them up or reduce memory consumption. if this turns out to be true, i can try contribute those upstream, but i suspect some optimisations will be too disruptive for the original repo to pull back in.

**Comment by sindresorhus** [comment]
It would be great if we could collaborate instead of duplicating efforts over minor differences. 🙏

Forks usually make sense when a project is abandoned or heading in a very different direction. Here, most of the changes seem like small preferences rather than fundamental shifts.

---

> empathic is faster and smaller than read-package-up (i hear you're moving to find-up-simple, but this will still be true)

It is done as of v59.

The difference between `empathic` and `find-up-simple` in size and speed is neglible in practice. [Please read](https://sindresorhus.com/blog/micro-benchmark-fallacy)

> is-builtin-module is unnecessary, node ships a built in list now (we have to exclude punycode since the npm package didn't include it)

It's useful because it provides a static list, so the result is consistent across Node.js versions: https://github.com/sindresorhus/is-builtin-module?tab=readme-ov-file#is-builtin-module

> indent-string is unnecessary, especially given it is used to apply a count of indentation but we always call it with 1 and set the indent character to the entire indent string (i.e. the already repeat-ed indent character)

Happy to have it removed.

> strip-indent can be replaced with dedent (will be about as big when it switches to ESM-only)

Unclear what the practical benefit would be?

> tinyexec de-dupes better than nano-spawn

It's just a development dependency, so that makes no practical difference. 40 kb is also negible in a non-browser context, and `nano-spawn` handles more edge-cases than `tinyexec`.

> a custom build of lodash is shipped pretty much just to get some string casing utils - replace this with scule, smaller and faster

Happy to do that as long as they cover all the edge-cases of the Lodash utilities.

> but i suspect some optimisations will be too disruptive for the original repo to pull back in

It would be great if you could expand on this. We are very much interested in improving performance.

*Keep in mind though that optimizations are only useful if they result in noticable performance difference in real-world projects, not micro-benchmarks.*

**Comment by 43081j** [comment]
> It is done as of v59.

see this quote from me:

> i hear you're moving to find-up-simple, but this will still be true

in case that wasn't written clearly enough, empathic is still faster and smaller so will remain the choice here.

> It's useful because it provides a static list, so the result is consistent across Node.js versions

i'm happy to align with the engine being used rather than an arbitrary list.

> Unclear what the practical benefit would be?

this is one of the smaller changes. both do roughly the same thing. `dedent` just happened to de-dupe better in some repos i tried out.

> It's just a development dependency, so that makes no practical difference. 40 kb is also negible in a non-browser context, and nano-spawn handles more edge-cases than tinyexec.

i'd like to improve DX for maintainers/contributors, too. so this was a drop-in no-brainer replacement for me. no edge cases nano-spawn apparently handles needed handling here it seems.

> Happy to do that as long as they cover all the edge-cases of the Lodash utilities

if somehow they didn't, the correct thing to do is contribute upstream to scule (or whatever utility has been chosen), not to hold on to lodash. but it seems everything works fine with scule anyway

> It would be great if you could expand on this. We are very much interested in improving performance

there's more investigation to be done, but i'm happy to try contribute any perf improvements i make here upstream.

> Keep in mind though that optimizations are only useful if they result in noticable performance difference in real-world projects, not micro-benchmarks

of course, i performance test a lot of projects, this wasn't in doubt.

**Comment by JounQin** [comment]
For `devDendpenciea`, I don't quite follow it could be a reason to fork personally. Every author has its own preferences, that's doesn't matter for end users, if there are no clear and huge benefits to migrate, I think the author's preferences are just fine.

For `dependencies`, we of course could and should find more adaptable dependencies. But for `is-builtin-module`, I don't think it's useful because it actually doesn't check the specific node version, for examine, `node:` protocol modules are not available on legacy Node versions, of course it could be argued that we don't support legacy Node versions, then why `module.isBuiltin` won't work? I still don't what does this package trying to resolve.

The most important part is, if it's forked, then which should I choose and why? This is really confusing for users.

**Comment by 43081j** [comment]
The devDependencies are not the reason to fork 👍 

Contributions should and will happen upstream, but it seems not all of them will be accepted. So a fork makes sense in that case.

I don't think we should sacrifice performance improvements, even if having two choices of lint plugin is more effort than having one. Choices should be fine, many lint plugins have alternatives already (while the original is maintained).

**Comment by fisker** [comment]
> why module.isBuiltin won't work?

Replace it with `module.isBuiltin()`, run unicorn against this code

```js
import x from 'inspector/promises'
```

On Node.js 18 & 20.

There will be more builtin modules added in future, static list makes result consistent.

**Comment by fisker** [comment]
Anyway, since `eslint-plugin-n` already have the same rule as we do, we already consider to remove `perfer-node-protocol` https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1778#issuecomment-2051236918

Currently unicorn version checkes more places than eslint-plugin-n, but once https://github.com/eslint-community/eslint-plugin-n/pull/430 land, they will be the same.

**Comment by JounQin** [comment]
> There will be more builtin modules added in future, static list makes result consistent.

That's why we should prefer using `module.isBuiltin`, Node 18 users shouldn't be able to consider Node 20 only modules as built-in, because it's just indeed unavailable.

**Comment by fisker** [comment]
Don't think so, people can lint with 18, run code in 20.

**Comment by JounQin** [comment]
> many lint plugins have alternatives already (while the original is maintained)

Useless there are huge improvements or new features, or if the DX is just terrible on upstream.

I don't see all od them in this case.

**Comment by JounQin** [comment]
> Don't think so, people can lint with 18, run code in 20.

It's just incorrect usage IMO, passing a module unavailable on Node 18 would result runtime error on Node 18, compatible module should be used instead of the latest.

**Comment by 43081j** [comment]
if you have code that depends on a module that only exists in 20.x, you shouldn't be trying to lint it in 18.x. clearly your code _requires_ 20.x.

if it ran in 18.x, it wouldn't be using modules only 20.x has.

**Comment by fisker** [comment]
You'll be running your lint on 22, and run code on 24, maybe next week.
Or maybe you already run lint on 22, and run code on 23.

**Comment by fisker** [comment]
And what Node.js version are you running in your VSCode?

**Comment by 43081j** [comment]
given `inspector/promises` was apparently added in 19.x.

any code that depends on that, _requires_ 19.x.

given that the code requires 19.x and above, it does not make sense to try lint the code in a lower version.

imagine having a repo which requires node 20.x but, for some reason, runs CI against 18.x (and always fails).

anyhow, this fork is not going to revert to using an npm package for something node itself provides.

**Comment by fisker** [comment]
It's just an example, static list is a more future-proof solution.


And why I can't write code like


```js
console.log(process.getBuiltinModule?.('inspector/promises')?.xxx() ?? 'fallback')
```

**Comment by 43081j** [comment]
we can agree to disagree.

i think we should be using the built in functionality node provides, and tying the list to the engine you're running in. we don't need to contribute that change upstream, it can remain here.

**Comment by sindresorhus** [comment]
> in case that wasn't written clearly enough, empathic is still faster and smaller so will remain the choice here.

I think you missed the second line of my comment:

> The difference between `empathic` and `find-up-simple` in size and speed is neglible in practice. [Please read](https://sindresorhus.com/blog/micro-benchmark-fallacy)

The micro-second performance difference has absoleutely no impact in `eslint-plugin-unicorn`. When it comes to performance, it's important to be practical and not get lost in micro-benchmarks. `eslint-plugin-unicorn` does not call the findup function thousands of times a second like a micro-benchmark does, it also caches the result.

In addition, the benchmark on `emphatic` is not even accurate: https://github.com/lukeed/empathic/issues/8

---

> i'd like to improve DX for maintainers/contributors, too.

I don't see how saving 40 kb of disk space improves DX in a meaningful way?

**Comment by JounQin** [comment]
`eslint.runtime: node` should definitely be used for `vscode-esint`: https://github.com/microsoft/vscode-eslint#:~:text=eslint.runtime,-%2D%20use%20this%20setting

**Comment by 43081j** [comment]
> I think you missed the second line of my comment:

i did not, and i did not get lost in or look at any micro-benchmarks. all good 👍 

empathic is smaller, de-dupes well, has a nicer API IMO. at the end of the day, anyone can implement the logic they both do, and it'll look mostly the same because none of this is difficult code, and that's why benchmarks don't tell us much. i chose empathic, you won't, that's fine.

> I don't see how saving 40 kb of disk space improves DX in a meaningful way?

the overall improvement of DX is meaningful, every little helps.

**Comment by fisker** [comment]
Speaking of something like find file up, I think the current implementation in unicorn should be improved, since it caches by file, while it should cache by directory and directories between the file and the found file. It's more effective.

When I wrote Prettier config search logic, https://github.com/prettier/prettier/blob/main/src/config/searcher.js I didn't see any lib already did this, I was planning to make one, but didn't get time to do https://www.npmjs.com/package/search-closest .

**Comment by JounQin** [comment]
I believe the biggest divergence is `is-builtin-module`, and it seems you two maintainers won't accept to removing it in favor of `module.isBuiltin`, maybe, we could mark `is-builtin-module` as optional peer dependency and we provide an option to enable this behavior, if the user desires the static list support, they can install it specifically.

**Comment by fisker** [comment]
As I said, we'll remove `perfer-node-protocol`, meaningless to discuss.

**Comment by 43081j** [comment]
that seems a good idea. we'll sync the fork when you do 👍 

once some of the performance changes land, maybe we can try contribute them upstream. though most will be large refactors i suspect, so may just live here.

**Comment by JounQin** [comment]
> As I said, we'll remove `perfer-node-protocol`, meaningless to discuss.

Well, good to know then.

**Comment by fisker** [comment]
I'm going to implement `search-closest`, see if I can beat both `empathic` and `find-up-simple`. 😄

**Comment by lukeed** [comment]
```
find-up-simple                 141.7 µs         7,059 (128.4 µs … 723.9 µs) 144.3 µs 161.0 µs 171.5 µs
find-up-simple (sync)           66.5 µs        15,050 ( 62.5 µs … 282.4 µs)  66.4 µs  74.2 µs  84.2 µs
find.up (sync)                   7.0 µs       143,300 (  6.9 µs …   7.0 µs)   7.0 µs   7.0 µs   7.0 µs
```

**Comment by JounQin** [comment]
@lukeed Wow, what's `find.up` by the way?

**Comment by lukeed** [comment]
`empathic/find`
